libsrc/apple2/ser/a2.ssc.s

```diff
@@ -210,6 +193,48 @@ SER_CLOSE:
         stx     Index           ; Mark port as closed
         rts
 
+;----------------------------------------------------------------------------
+; Internal function to test whether we're really talking with a 6551,
+; or at least it really looks like it.
+; Input: X= Index
+; Output: carry set if this is very probably a 6551, clear if not
+
+Verify6551:
+        lda     ACIA_STATUS,x   ; Save current values in what we expect to be
+        pha                     ; the ACIA status register
+        lda     ACIA_CMD,x      ; and command register. So we can restore them
+        sta     tmp1            ; if this isn't a 6551.
+
+        and     #$01
+        bne     NotAcia         ; We expect command register bit 0 to be 0
+
+        lda     tmp1
+        ora     #$01            ; Enable receiver/transmitter
+        sta     ACIA_CMD,x
+        sta     tmp2            ; Store it for comparison
+
+        lda     ACIA_CMD,x      ; Is command register what we wrote?
+        cmp     tmp2
+        bne     NotAcia
+
+        sta     ACIA_STATUS,x   ; Reset Acia (value written is not important)
+
+        lda     ACIA_CMD,x      ; Is the value back to the original value?
+        cmp     tmp1
+        bne     NotAcia
+
+        pla                     ; Clear stack of leftovers
+        clc
+        rts
+
+NotAcia:
+        lda     tmp1            ; Restore saved bytes, they're probably needed
+        sta     ACIA_CMD,x
+        pla
+        sta     ACIA_STATUS,x
+        sec
+        rts
```

Comment:
I'd suggest to restructure this function a bit...

* After line 222, you can do an LSR to directly have the desired information in the carry flag.
* You can just always execute the restore of the saved values in lines 231-234 as we're not optimizing for cycles, but for bytes. Those instructions all don't change the carry flag.
* As far as I can see, when taking my comment above into account, it's shorter (and clearer) to save both original values on the stack.